### PR TITLE
feat: only use spring `@Nullable` annotation in the Spring Data Repository class

### LINKED ISSUE #2: Check Nullable annotation usage in Spring Data Repository classes
Spring Data Repositories only support the usage of their own `org.springframework.lang.Nullable` annotations[^1]. Misuses could lead to runtime errors when invoking repository methods and thus should be prevented.

[^1]: https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.nullability.annotations

### PR DESCRIPTION
**Pull request by Charm-Xin**
close https://github.com/ocpddev/jsensible/issues/2

### CODE REVIEW

File: src/main/kotlin/dev/ocpd/jsensible/internal/jpa/NotSpringNullable.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+     * Matches not spring nullable.
+     */
+    internal fun noUseSpringNullable(): ArchCondition<in JavaClass> =
+        dependOnClassesThat(
+            assignableTo(
+                nullableClasses()
+                    .and(not(springNullableClasses()))
+            )
+        ).`as`("no use spring nullable")
+}
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/internal/jpa/NotSpringNullable.kt]: Missing empty line here.

File: src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt [hunk 2]
```diff
@@ -25,4 +27,12 @@ internal object NullabilityAnnotations {
             "org.springframework.lang.NonNull"
         )
     ).forSubtype()
+
+    internal fun nullableClasses(): DescribedPredicate<JavaClass> =
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt]:
```suggestion
    internal fun nullableAnnotations(): DescribedPredicate<JavaClass> =
```

File: src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt [hunk 3]
```diff
@@ -25,4 +27,12 @@ internal object NullabilityAnnotations {
...
+        simpleName("Nullable")
+
+    internal fun springNullableClasses(): DescribedPredicate<JavaClass> =
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt]:
```suggestion
    internal fun springNullableAnnotation(): DescribedPredicate<JavaClass> =
```

File: src/main/kotlin/dev/ocpd/jsensible/rules/jpa/JpaRules.kt [hunk 4]
```diff
@@ -29,4 +32,19 @@ object JpaRules {
...
+    /**
+     * only support the [org.springframework.lang.Nullable] annotation
+     *
+     * Spring Data Repositories only support the usage of their own
+     * [org.springframework.lang.Nullable] annotation. Misuses could
+     * lead to runtime errors when invoking repository methods.
+     *
+     * Solution: Use the [org.springframework.lang.Nullable] annotation.
+     */
+    fun useSpringNullableAnnotation(): ArchRule =
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/rules/jpa/JpaRules.kt]:
```suggestion
    fun onlyUseNullableAnnotationFromSpring(): ArchRule =
```

File: src/main/kotlin/dev/ocpd/jsensible/internal/jpa/NotSpringNullable.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * No use spring nullable conditions.
+ */
+internal object NotSpringNullable {
+
+    /**
+     * Matches not spring nullable.
+     */
+    internal fun noUseSpringNullable(): ArchCondition<in JavaClass> =
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/internal/jpa/NotSpringNullable.kt]:
```suggestion
    internal fun notUseSpringNullable(): ArchCondition<in JavaClass> =
```

File: src/main/kotlin/dev/ocpd/jsensible/internal/jpa/NotSpringNullable.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    /**
+     * Matches not spring nullable.
+     */
+    internal fun notUseSpringNullable(): ArchCondition<in JavaClass> =
+        dependOnClassesThat(
+            assignableTo(
+                nullableAnnotations()
+                    .and(not(springNullableAnnotation()))
+            )
+        ).`as`("no use spring nullable")
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/internal/jpa/NotSpringNullable.kt]:
```suggestion
        ).`as`("not use spring nullable")
```

File: src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt [hunk 7]
```diff
@@ -10,19 +12,19 @@ import dev.ocpd.jsensible.internal.JavaClassPredicates.anyOf
...
             "org.jetbrains.annotations.NotNull",
             "javax.annotation.Nonnull",
             "org.springframework.lang.NonNull"
         )
     ).forSubtype()
+
+    internal fun nullableAnnotations(): DescribedPredicate<JavaClass> =
+        simpleName("Nullable")
+
+    internal fun springNullableAnnotation(): DescribedPredicate<JavaClass> =
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt:27]: Just curious. It's necessary to wrap this into a `anyOf()`?

File: src/main/kotlin/dev/ocpd/jsensible/rules/jpa/JpaRules.kt [hunk 8]
```diff
@@ -29,4 +32,19 @@ object JpaRules {
         noMembers().should(useEagerFetch())
             .`as`("use eager fetch")
             .because("no property should be fetched eagerly by default")
+
+    /**
+     * only support the [org.springframework.lang.Nullable] annotation
```
Comment (EggOxygen) [inline src/main/kotlin/dev/ocpd/jsensible/rules/jpa/JpaRules.kt]:
```suggestion
     * Only support the [org.springframework.lang.Nullable] annotation
```

**Review by EggOxygen — CHANGES_REQUESTED** [review]

File: src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt (hunk 7 above)
Comment (Charm-Xin) [reply src/main/kotlin/dev/ocpd/jsensible/internal/nullability/NullabilityAnnotations.kt:27]: After my research, I found that it is not necessary and the modification has been tested without problems.

**Review by EggOxygen — APPROVED** [review]

**Comment by unlimitedsola** [comment]
This should belong to the `spring` package as the Spring repositories are not limited to only JPA.

**Pull request closed without merging**